### PR TITLE
Refactor  codebase, implementing defense feedback

### PR DESCRIPTION
**Pull request by aknwosu**
This pull request makes changes to this codebase in action feedback received after the checkpoint defense

Improvements;

* Fixed the gulp function
* Adhered to Airbnb specs
* Test cases were made more specific
* Refactored the Search Index method to the set standard
* Adhered to ES6 standards
* Factored book "title" in the create index method
* Displayed messages to the front end using status messages
* Refactored the try-catch method

### CODE REVIEW

File: jasmine/spec/bundle.js [hunk 1]
```diff
@@ -3,64 +3,76 @@
 const emptyFile = require('./test/emptyFile.json');
```
Comment (andela-kakpobome) [inline jasmine/spec/bundle.js]: ### Critical

Files that are generated from other files should not be added to version control, eg a bundled file

File: jasmine/spec/bundle.js [hunk 2]
```diff
@@ -3,64 +3,76 @@
...
 const wrongFile = require('./test/wrongJsonObject.json');
 const validBook = require('./test/books.json');
-const trollTest = require('./test/trollTest.json');
 
 const invertedIndexTest = new InvertedIndexClass();
 
 describe('Read book data', () => {
-  it('reads a JSON file and asserts that it not empty', () => {
-    expect(invertedIndexTest.validFiles(emptyFile)[1]).toBe('Error, empty file');
+  it('reads a JSON file and asserts that it not empty',
```
Comment (andela-kakpobome) [inline jasmine/spec/bundle.js]: ### Nitpick

This should be `it's not empty`

File: jasmine/spec/bundle.js [hunk 3]
```diff
@@ -3,64 +3,76 @@
...
+  () => {
+    expect(invertedIndexTest.validFiles(emptyFile)[1])
```
Comment (andela-kakpobome) [inline jasmine/spec/bundle.js]: ### Critical

I already gave you feedback on this to return an object shaped like so:

```
{
  "status": true,
  "statusMessage": "Some message"
}
```
That'll make your code more readable. Then you'ld have something like, `.validFiles(emptyFile).status.toBe('Error, empty file')`

File: jasmine/spec/bundle.js [hunk 4]
```diff
@@ -3,64 +3,76 @@
...
+  () => {
+    expect(invertedIndexTest.createIndex('booky.json', validBook)['booky.json']
+    .alice).toBeDefined();
   });
 
-  it('should map the string keys to the correct objects in the JSON array', () => {
-    expect(invertedIndexTest.createIndex('books.json', validBook)['books.json'].of).toEqual({ 0: true, 1: true });
+  it('should map the string keys to the correct objects in the JSON array',
+  () => {
+    expect(invertedIndexTest.createIndex('books.json', validBook)['books.json']
```
Comment (andela-kakpobome) [inline jasmine/spec/bundle.js]: ### Critical

Create Index should not return anything (maybe except a Boolean), this should instead set an internal variable that `getIndex` returns when called.

File: jasmine/spec/bundle.js [hunk 5]
```diff
@@ -3,64 +3,76 @@
...
+    .of).toEqual({ 0: true, 1: true });
   });
+
   it('Should confirm correct index is returned', () => {
-    expect(invertedIndexTest.getIndex()['booky.json'].of).toEqual({ 0: true, 1: true });
-    expect(Object.keys(invertedIndexTest.getIndex()['booky.json']).length).toEqual(25);
+    expect(invertedIndexTest.getIndex()['booky.json'].of)
```
Comment (andela-kakpobome) [inline jasmine/spec/bundle.js]: ### Critical

`getIndex` when passed a filename, should return the index for that file. That'ld be preferable to what you've done(return the entire index before plucking it out).

File: jasmine/spec/bundle.js [hunk 6]
```diff
@@ -3,64 +3,76 @@
...
+    .toEqual({ 0: true, 1: true });
+    expect(Object.keys(invertedIndexTest.getIndex()['booky.json'])
+    .length).toEqual(31);
   });
 });
 
 describe('Search index', () => {
-  it('returns the indices of the correct objects that contain the words in the search query', () => {
-    expect(invertedIndexTest.searchIndex('Alice Bicycle', 'booky.json')).toEqual({ 'booky.json': { alice: { 0: true } } });
+  it('returns the correct objects indices that contain the searched words',
```
Comment (andela-kakpobome) [inline jasmine/spec/bundle.js]: ### Nitpick

Should be `that contains`

File: src/inverted-index.js [hunk 7]
```diff
@@ -6,99 +6,110 @@
...
-    for (let i = 0; i < books.length; i += 1) {
-      const textIndex = books[i].text.toLowerCase().match(/[\w']+/g);
-      for (let j = 0; j < textIndex.length; j += 1) {
-        if (this.index[filename][textIndex[j]] === undefined) {
-          this.index[filename][textIndex[j]] = {};
-          this.index[filename][textIndex[j]][i] = true;
+    this.books = (books);
+    let bookIndex;
+    books.forEach((singleBook, inex) => {
+      // console.log(singleBook);
```
Comment (andela-kakpobome) [inline src/inverted-index.js]: ### Critical

Why is this here?

File: src/inverted-index.js [hunk 8]
```diff
@@ -6,99 +6,110 @@
...
-
-    for (let i = 0; i < books.length; i += 1) {
-      const textIndex = books[i].text.toLowerCase().match(/[\w']+/g);
-      for (let j = 0; j < textIndex.length; j += 1) {
-        if (this.index[filename][textIndex[j]] === undefined) {
-          this.index[filename][textIndex[j]] = {};
-          this.index[filename][textIndex[j]][i] = true;
+    this.books = (books);
+    let bookIndex;
+    books.forEach((singleBook, inex) => {
```
Comment (andela-kakpobome) [inline src/inverted-index.js]: ### Critical

Please use reasonable variable names, `inex` does not make sense.

File: src/inverted-index.js [hunk 9]
```diff
@@ -6,99 +6,110 @@
...
+    });
     return this.index;
   }
+
   /**
-     * returns created index
-     * 
-     * @returns{object};
-     */
+   * returns created index
```
Comment (andela-kakpobome) [inline src/inverted-index.js]: ### Nitpick

The rest of the doc blocks have the name of the method, why is this one exempt?

File: src/inverted-index.js [hunk 10]
```diff
@@ -6,99 +6,110 @@
...
-        queries = queries.join();
+      if (typeof (terms) === typeof []) {
+        terms = terms.join();
       }
-      queries = queries.toLowerCase().match(/\w+/g);
-      for (let word = 0; word < queries.length; word += 1) {
-        if (this.index[searchKey][queries[word]]) {
-          searchResult[searchKey][queries[word]] = this.index[searchKey][queries[word]];
+      terms = terms.toLowerCase().match(/[\w']+/g);
+      for (let word = 0; word < terms.length; word += 1) {
```
Comment (andela-kakpobome) [inline src/inverted-index.js]: ### Critical

Stick with the `forEach` API instead.

**Review by andela-kakpobome — CHANGES_REQUESTED** [review]
Fix all these issues.

**Comment by andela-kakpobome** [comment]
@andela-anwosu your [Angular code base](https://github.com/andela-anwosu/inverted-index-cp1/blob/29ca913f887d360367230995084063f5a8e8c71e/controllers/app.js) is also riddled with `console.logs`, `alerts` and `// commented out code`.

Rethink and refactor accordingly.

File: src/inverted-index.js (hunk 10 above)
Comment (aknwosu) [reply src/inverted-index.js]: fixed

File: jasmine/spec/bundle.js (hunk 1 above)
Comment (aknwosu) [reply jasmine/spec/bundle.js]: fixed

File: jasmine/spec/bundle.js (hunk 2 above)
Comment (aknwosu) [reply jasmine/spec/bundle.js]: fixed

File: jasmine/spec/bundle.js (hunk 3 above)
Comment (aknwosu) [reply jasmine/spec/bundle.js]: fixed

File: jasmine/spec/bundle.js (hunk 4 above)
Comment (aknwosu) [reply jasmine/spec/bundle.js]: fixed

File: jasmine/spec/bundle.js (hunk 5 above)
Comment (aknwosu) [reply jasmine/spec/bundle.js]: fixed

File: jasmine/spec/bundle.js (hunk 6 above)
Comment (aknwosu) [reply jasmine/spec/bundle.js]: completed

File: src/inverted-index.js (hunk 7 above)
Comment (aknwosu) [reply src/inverted-index.js]: fixed

File: src/inverted-index.js (hunk 8 above)
Comment (aknwosu) [reply src/inverted-index.js]: fixed. Now bookIndex

File: src/inverted-index.js (hunk 9 above)
Comment (aknwosu) [reply src/inverted-index.js]: fixed

File: controllers/app.js [hunk 11]
```diff
@@ -3,78 +3,78 @@
...
-        // $scope.newIndex = new InvertedIndexClass();
-        const reader = new FileReader();
-        reader.readAsText(fileDoc);
-        reader.onload = (event) => {
-            $scope.readybooks = event.target.result;
-            $scope.validationResult = $scope.newIndex.validFiles($scope.readybooks);
+  $scope.uploader = () => {
+    const fileDoc = document.getElementById('myJsonFile').files[0];
+    $scope.fileDocName = fileDoc.name;
+    if (typeof (fileDoc !== 'Blob')) {
```
[the remark this answers is not in the document]
Comment (andela-kakpobome) [reply controllers/app.js]: ### Critical

The `typeof` here serves absolutely no purpose, why is it here?

File: controllers/app.js [hunk 12]
```diff
@@ -3,78 +3,78 @@
...
-            $scope.titles[$scope.indexToBeCreated] = titlesList($scope.newIndex.books.length);
-            $scope.arrayLength[$scope.indexToBeCreated] = indexCount($scope.newIndex.books.length);
-            $scope.showIndex = true;
-            $scope.showSearch = false;
+      $scope.$apply(() => {
+        if ($scope.validationResult.status === true) {
+          $scope.allValidatedBooks[$scope.fileDocName] = $scope.readybooks;
+          $scope.uploaderError = 'success! Please create an index of the uploaded book below';
+        } else {
+          alert($scope.validationResult.statusMessage); // eslint-disable-line no-alert
```
Comment (andela-kakpobome) [inline controllers/app.js]: ### Critical

There are still alerts, why???????!!!!!!

File: controllers/app.js [hunk 13]
```diff
@@ -3,78 +3,78 @@
...
+      $scope.showSearch = false;
+    }
+    $scope.allTitles = $scope.titles;
+    $scope.allArrayLength = $scope.arrayLength;
+  };
 
 
-    $scope.searchJson = () => {
+  $scope.searchJson = () => {
         // let file = $scope.indexToBeSearched || null,
```
Comment (andela-kakpobome) [inline controllers/app.js]: ### Critical

@andela-anwosu you are still really careless with things like this (commented out code), why?!

File: controllers/app.js [hunk 14]
```diff
@@ -3,78 +3,78 @@
...
-        }
-    };
-});
\ No newline at end of file
+    if ($scope.searchTerms === undefined) {
+      $scope.showError = 'Please enter valid search terms';
+    } else if ($scope.indexOfWords === undefined) {
+      $scope.showError = 'Please upload a Json object first before searching';
+    } else {
+      if ($scope.indexToBeSearched === undefined) {
```
Comment (andela-kakpobome) [inline controllers/app.js]: ### Critical

Why is this not on the same line with `else`? Use an `else if`

File: jasmine/spec/inverted-index-test.js [hunk 15]
```diff
@@ -2,59 +2,71 @@
...
+  () => {
+    expect(invertedIndexTest.getIndex('books.json')).toBeDefined();
   });
-  it('Should confirm correct index is returned', () => {
-    expect(invertedIndexTest.getIndex()['booky.json'].of).toEqual({ 0: true, 1: true });
-    expect(Object.keys(invertedIndexTest.getIndex()['booky.json']).length).toEqual(25);
+
+  it('should map the string keys to the correct objects in the JSON array',
+  () => {
+    expect(invertedIndexTest.getIndex()['books.json']
```
Comment (andela-kakpobome) [inline jasmine/spec/inverted-index-test.js]: ### Critical

You only corrected `.getIndex()['books.json']` in one instance, you should pass `books.json` as an argument!

**Review by andela-kakpobome — CHANGES_REQUESTED** [review]
@andela-anwosu yet again, fix these issues!

File: controllers/app.js (hunk 12 above)
Comment (aknwosu) [reply controllers/app.js]: fixed. I mentioned it yesterday that the errors were for just error messages. Taken them out now and they display just beneath the upload button.

File: controllers/app.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (aknwosu) [reply controllers/app.js]: fixed

**Pull request closed without merging**